packages/button/src/Button/Button.tsx

```diff
@@ -0,0 +1,184 @@
+import { Icon as TelegraphIcon } from "@telegraph/icon";
+import { Text as TypographyText } from "@telegraph/typography";
+import clsx from "clsx";
+import React from "react";
+
+import {
+  buttonColorMap,
+  buttonSizeMap,
+  iconColorMap,
+  iconSizeMap,
+  iconVariantMap,
+  textColorMap,
+  textSizeMap,
+} from "./Button.constants";
+
+type RootBaseProps = {
+  variant?: "solid" | "soft" | "outline" | "ghost";
+  size?: "1" | "2" | "3";
+  color?: "accent" | "gray" | "red";
+  state?: "default" | "loading" | "disabled" | "error" | "success" | "warning";
+};
+
+type InternalProps = {
+  layout: "default" | "icon-only";
+};
+
+type RootProps = RootBaseProps & React.ButtonHTMLAttributes<HTMLButtonElement>;
+
+type RootRef = HTMLButtonElement;
+
+// Might make sense to move this into a shared
+// package to be used across all components if we
+// start to see more of these kinds of types.
+type Required<T> = {
+  [P in keyof T]-?: T[P];
+};
+
+const ButtonContext = React.createContext<
+  Required<RootBaseProps & InternalProps>
+>({
+  variant: "solid",
+  size: "2",
+  color: "gray",
+  state: "default",
+  layout: "default",
+});
+
+const Root = React.forwardRef<RootRef, RootProps>(
+  (
+    {
+      variant = "solid",
+      size = "2",
+      color = "gray",
+      state: initialState = "default",
+      disabled,
+      className,
+      ...props
+    },
+    forwardedRef,
+  ) => {
+    const state = disabled ? "disabled" : initialState;
+    const [layout, setLayout] =
+      React.useState<InternalProps["layout"]>("default");
+
+    // Check if "Icon" is the only child so we can set the kind of layout
+    React.useEffect(() => {
+      const children = React.Children.toArray(props.children);
+      if (children?.length === 1 && React.isValidElement(children[0])) {
+        const child = children[0] as React.ReactComponentElement<typeof Icon>;
+        if (child?.type?.displayName === "Icon") {
+          return setLayout("icon-only");
+        }
+        return setLayout("default");
+      }
+    }, [props.children]);
```

Comment:
Open to feedback on a better way to do this. Could do it with css but don't want to muddy up how the tokens are setup for size. This also throws a warning in RSC as the className doesn't match first render. Maybe a ref could be good?

Reply:
Yeah this doesn't feel super clean, but based on the current design system it seems like the choice is to either handle the "icon only" styles with one off CSS, or with this. i.e. no easy way to avoid that particular complexity, and at the end of the day it's not very consequential as long as it renders the right thing, given it's contained in this component. 

I guess my only other thought/question — do we need to do this with `useState`+`useEffect` though? This is the only place where this state is set, which makes me think we can just use `useMemo`?

Reply:
Yeah I'm going to take another stab at removing the useEffect in exchange for something else